### ISSUE TITLE
draft-06 Support

### ISSUE THREAD

**Issue by khakulov** [Julian/jsonschema#337]
Is it in plan to support draft-06?

in my Project i need to use `propertyNames` to validate object keys agains regex pattern.
but `propertyNames` was introduced in draft-06.

for now i am using `patternProperties` with `"additionalProperties": False` together, but the error is confusing because of "schema does not allow additional properties"

**Comment by Julian** [comment]
It's definitely planned yep, just a bit behind. Patches definitely welcome.

**Comment by bsmithers** [comment]
I've just raised a PR (https://github.com/Julian/jsonschema/pull/370) addressing the currently failing tests. 

To continue the conversation in the appropriate place, are there particular components of the above checklist that you'd appreciate some help with? Or indeed anything new you have realised that needs work before draft6 support is complete?

**Comment by fireundubh** [comment]
FYI: Draft 7 was recently published.

**Comment by quantum13** [comment]
Hi, what tasks from https://github.com/Julian/jsonschema/issues/337#issuecomment-305981069 are still undone?

**Comment by Julian** [comment]
@quantum13 so the main one is to update the draft-06 branch with @bsmithers' type checker changes. I was hoping/planning to get a release out first just containing those, and then to get to merging them into the draft 6 stuff.

Out of curiosity for anyone following this ticket, do y'all have interest specifically in draft 6, or would jumping directly to draft 7/8 be just as acceptable?

**Comment by Julian** [comment]
I think https://github.com/python-hyper/uritemplate/pull/36 is still open as well, so we'd need to help out with getting that one merged there.

**Comment by adamcharnock** [comment]
As a follower of this ticket: @Julian, I'd be happy to jump to 7/8.

**Comment by Julian** [comment]
@adamcharnock thanks.

(Want to by the way say to all those following this ticket that I wish these things would get done faster but unfortunately open source gets an exceedingly small fraction of my time these days with all of the other fun things in life... Really am hoping to get out a release with newer draft support within the next 2-3 weeks though.)

**Comment by handrews** [comment]
> As a follower of this ticket: @Julian, I'd be happy to jump to 7/8.

@adamchainz draft-07 is a pretty small incremental step that is backwards-compatible with draft-06, so it should be relatively fast to add, and near-trivial to support both.  draft-08 will be more substantial, so it would be really good to at least get to draft-06 (and past draft-04) rather than wait for the even larger draft-08.

**Comment by adamchainz** [comment]
wrong adam

**Comment by iddan** [comment]
I would be happy to jump right ahead to 7/8. If there is no impl for draft-06 in python how many people would actually miss it? I mean: I assume the majority of the users probably would just want to upgrade to the latest version.

**Comment by mxmader** [comment]
> I assume the majority of the users probably would just want to upgrade to the latest version.
In my case the response is "if and only if there are no _breaking_ changes"

**Comment by handrews** [comment]
@iddan draft-06 has had more implementations in the wild for quiet some time now, in JavaScript, Java, .NET, Go, PHP, and perhaps others.  It is important to support interoperability while the drafts are evolving.

Plus, I will repeat, **again**:

draft-07 is a small increment on top of draft-06.  If you implement draft-07 you already implemented draft-06.  Since draft-06 is *smaller*, just implement that much, release it, then finish draft-07.

Please everyone stop making this a more complicated decision than it needs to be.

**Comment by Julian** [comment]
OK, still a few more outstanding items here, but making some progress...

Anyone who wants an early look is welcome to try out the `draft-06` branch -- comments obviously welcome.

There likely will be a beta in a few days as soon as some more of the outstanding checklist is checked off, but I believe things should at least function at the minute.

**Comment by julesjulian** [comment]
I think you meant to tag @Julian 😉

**Comment by graingert** [comment]
@julesjulian fixed!

**Comment by Julian** [comment]
@graingert yeah I should... will have a look at cutting one hopefully this weekend

**Comment by graingert** [comment]
@Julian it would be super awesome to get travis to automatically ship every commit as: with the dev version and commit hash: `jsonschema 2.6.1.dev237+g9632422`

**Comment by Julian** [comment]
If you send a PR that does that I'm happy to merge it!

**Comment by graingert** [comment]
@Julian looks like most of the work is down with `use_scm_version`, this just triggers the dpeloy. 
https://github.com/Julian/jsonschema/pull/456

The last bit, you have to do: is add deploy credentials - I'd recommend using a new account with maintainership access only

**Comment by Julian** [comment]
@graingert I don't think PyPI lets you upload dev releases so not sure it'd work completely as is but will have a look at your PR, thanks!

**Comment by graingert** [comment]
I'm pretty sure it does allow dev releases...

**Comment by Julian** [comment]
```
⊙  ~[jsonschema:python] setup.py bdist_wheel sdist upload >/dev/null                                                                                             Julian@Magbook ●
warning: build_py: byte-compiling is disabled, skipping.

warning: install_lib: byte-compiling is disabled, skipping.

Password:
Upload failed (400): '3.0.0a3.dev74+g3eabefb' is an invalid value for Version. Error: Can't use PEP 440 local versions. See https://packaging.python.org/specifications/core-metadata
error: Upload failed (400): '3.0.0a3.dev74+g3eabefb' is an invalid value for Version. Error: Can't use PEP 440 local versions. See https://packaging.python.org/specifications/core-metadata
```

**Comment by iddan** [comment]
When importing version 2.6.0 I don't see Draft6Validator. Does that make sense?

**Comment by Julian** [comment]
6 support is in the alpha releases only (3.0+)
